### ISSUE TITLE
Main and renderer processes should communicate with each other through actions and effects, not plain events

### ISSUE THREAD

**Issue by victorporof** [mozilla/tofino#1414]
tl;dr: go full stack sagas.

We're relying on ipc communication between the main process and the renderer process quite heavily. In both places, where events are handled, certain operations are described imperatively and in place. This makes testing difficult and application logic scattered. I'll attempt to describe a potentially better way of handling everything.

Currently, the renderer process sends events to the main process:
_SCENARIO 1_: whenever it thinks the main process requires some bits and pieces of the application state (pinned tabs, currently focused element etc.), or
_SCENARIO 2_: when some actions need to happen that need to be handled in the main process (opening the application menu, persisting application state for session restore etc).

(Opening the application menu doesn't *have* to be handled in the main process and should live in the frontend. Issue https://github.com/mozilla/tofino/issues/1413 is filed to fix that)

On the other hand, the main process sends events to the renderer process:
_SCENARIO 3_: whenever some actions need to happen in the renderer process (see all the event handlers [here](https://github.com/mozilla/tofino/blob/master/app/ui/browser-modern/setup-ipc.js))

All of the operations described above are essentially impure side effects: by definition, none of them are caused by anything in the current process. The fact that they might lead (and they do lead) to pure operations is besides the point; our side effects handled in sagas always lead to pure operations.

Stepping back, all of this looks remarkably similar to the current (localized) way of handling effects in the renderer process alone: "something" causes some side effects to happen, which potentially lead to pure operations. All of the ipc communication can be described in the exactly the same way: an ipc message is "something" that causes a side effect to happen, which potentially leads to pure operations.

We therefore have a very elegant (and extremely easily testable) solution for handling all of this with full-stack redux-sagas.

After https://github.com/mozilla/tofino/issues/1354 is fixed, the renderer process and the main process will both have redux stores. The renderer process will send *actions*, not *events* via ipc to update main process's redux store, taking care of _SCENARIO 1_ described above. We should go all the way and handle _SCENARIO 2_ and _SCENARIO 3_ in exactly the same way, turning the current events into actions, and gather the scattered application logic into sagas.

Stepping back again, this is *exactly* how we're handling messages from the user agent service after the browser-modern frontend was written: all messages from the UAS are actions, which are simply dispatched to the redux store, and the existing infrastructure handles them elegantly.

By fixing this issue, we'll unify everything.
